File: utils/plugins/lspRecommendation.ts
/**
 * LSP Plugin Recommendation Utility
 *
 * Scans installed marketplaces for LSP plugins and recommends plugins
 * based on file extensions, but ONLY when the LSP binary is already
 * installed on the system.
 *
 * Limitation: Can only detect LSP plugins that declare their servers
 * inline in the marketplace entry. Plugins with separate .lsp.json files
 * are not detectable until after installation.
 */

import { extname } from 'path'
import { randomUUID } from 'crypto'
import { isBinaryInstalled } from '../binaryCheck.js'
import { getGlobalConfig, saveGlobalConfig } from '../config.js'
import { logForDebugging } from '../debug.js'
import { isPluginInstalled } from './installedPluginsManager.js'
import { installResolvedPlugin } from './pluginInstallationHelpers.js'
import {
  getMarketplace,
  getPluginById,
  loadKnownMarketplacesConfig,
} from './marketplaceManager.js'
import {
  ALLOWED_OFFICIAL_MARKETPLACE_NAMES,
  type PluginMarketplaceEntry,
} from './schemas.js'

/**
 * LSP plugin recommendation returned to the caller
 */
export type LspPluginRecommendation = {
  pluginId: string // "plugin-name@marketplace-name"
  pluginName: string // Human-readable plugin name
  marketplaceName: string // Marketplace name
  description?: string // Plugin description
  isOfficial: boolean // From official marketplace?
  extensions: string[] // File extensions this plugin supports
  command: string // LSP server command (e.g., "typescript-language-server")
}

// Maximum number of times user can ignore recommendations before we stop showing
const MAX_IGNORED_COUNT = 5

/**
 * Check if a marketplace is official (from Mossen)
 */
function isOfficialMarketplace(name: string): boolean {
  return ALLOWED_OFFICIAL_MARKETPLACE_NAMES.has(name.toLowerCase())
}

/**
 * Internal type for LSP info extracted from plugin manifest
 */
type LspInfo = {
  extensions: Set<string>
  command: string
}

/**
 * Extract LSP info (extensions and command) from inline lspServers config.
 *
 * NOTE: Can only read inline configs, not external .lsp.json files.
 * String paths are skipped as they reference files only available after installation.
 *
 * @param lspServers - The lspServers field from PluginMarketplaceEntry
 * @returns LSP info with extensions and command, or null if not extractable
 */
function extractLspInfoFromManifest(
  lspServers: PluginMarketplaceEntry['lspServers'],
): LspInfo | null {
  if (!lspServers) {
    return null
  }

  // If it's a string path (e.g., "./.lsp.json"), we can't read it from marketplace
  if (typeof lspServers === 'string') {
    logForDebugging(
      '[lspRecommendation] Skipping string path lspServers (not readable from marketplace)',
    )
    return null
  }

  // If it's an array, process each element
  if (Array.isArray(lspServers)) {
    for (const item of lspServers) {
      // Skip string paths in arrays
      if (typeof item === 'string') {
        continue
      }
      // Try to extract from inline config object
      const info = extractFromServerConfigRecord(item)
      if (info) {
        return info
      }
    }
    return null
  }

  // It's an inline config object: Record<string, LspServerConfig>
  return extractFromServerConfigRecord(lspServers)
}

/**
 * Extract LSP info from a server config record (inline object format)
 */
/**
 * Type guard to check if a value is a record object
 */
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

function extractFromServerConfigRecord(
  serverConfigs: Record<string, unknown>,
): LspInfo | null {
  const extensions = new Set<string>()
  let command: string | null = null

  for (const [_serverName, config] of Object.entries(serverConfigs)) {
    if (!isRecord(config)) {
      continue
    }

    // Get command from first valid server config
    if (!command && typeof config.command === 'string') {
      command = config.command
    }

    // Collect all extensions from extensionToLanguage mapping
    const extMapping = config.extensionToLanguage
    if (isRecord(extMapping)) {
      for (const ext of Object.keys(extMapping)) {
        extensions.add(ext.toLowerCase())
      }
    }
  }

  if (!command || extensions.size === 0) {
    return null
  }

  return { extensions, command }
}

/**
 * Internal type for plugin with LSP info
 */
type LspPluginInfo = {
  entry: PluginMarketplaceEntry
  marketplaceName: string
  extensions: Set<string>
  command: string
  isOfficial: boolean
}

/**
 * Get all LSP plugins from all installed marketplaces
 *
 * @returns Map of pluginId to plugin info with LSP metadata
 */
async function getLspPluginsFromMarketplaces(): Promise<
  Map<string, LspPluginInfo>
> {
  const result = new Map<string, LspPluginInfo>()

  try {
    const config = await loadKnownMarketplacesConfig()

    for (const marketplaceName of Object.keys(config)) {
      try {
        const marketplace = await getMarketplace(marketplaceName)
        const isOfficial = isOfficialMarketplace(marketplaceName)

        for (const entry of marketplace.plugins) {
          // Skip plugins without lspServers
          if (!entry.lspServers) {
            continue
          }

          const lspInfo = extractLspInfoFromManifest(entry.lspServers)
          if (!lspInfo) {
            continue
          }

          const pluginId = `${entry.name}@${marketplaceName}`
          result.set(pluginId, {
            entry,
            marketplaceName,
            extensions: lspInfo.extensions,
            command: lspInfo.command,
            isOfficial,
          })
        }
      } catch (error) {
        logForDebugging(
          `[lspRecommendation] Failed to load marketplace ${marketplaceName}: ${error}`,
        )
      }
    }
  } catch (error) {
    logForDebugging(
      `[lspRecommendation] Failed to load marketplaces config: ${error}`,
    )
  }

  return result
}

/**
 * Find matching LSP plugins for a file path.
 *
 * Returns recommendations for plugins that:
 * 1. Support the file's extension
 * 2. Have their LSP binary installed on the system
 * 3. Are not already installed
 * 4. Are not in the user's "never suggest" list
 *
 * Results are sorted with official marketplace plugins first.
 *
 * @param filePath - Path to the file to find LSP plugins for
 * @returns Array of matching plugin recommendations (empty if none or disabled)
 */
export async function getMatchingLspPlugins(
  filePath: string,
): Promise<LspPluginRecommendation[]> {
  // Check if globally disabled
  if (isLspRecommendationsDisabled()) {
    logForDebugging('[lspRecommendation] Recommendations are disabled')
    return []
  }

  // Extract file extension
  const ext = extname(filePath).toLowerCase()
  if (!ext) {
    logForDebugging('[lspRecommendation] No file extension found')
    return []
  }

  logForDebugging(`[lspRecommendation] Looking for LSP plugins for ${ext}`)

  // Get all LSP plugins from marketplaces
  const allLspPlugins = await getLspPluginsFromMarketplaces()

  // Get config for filtering
  const config = getGlobalConfig()
  const neverPlugins = config.lspRecommendationNeverPlugins ?? []

  // Filter to matching plugins
  const matchingPlugins: Array<{ info: LspPluginInfo; pluginId: string }> = []

  for (const [pluginId, info] of allLspPlugins) {
    // Check extension match
    if (!info.extensions.has(ext)) {
      continue
    }

    // Filter: not in "never" list
    if (neverPlugins.includes(pluginId)) {
      logForDebugging(
        `[lspRecommendation] Skipping ${pluginId} (in never suggest list)`,
      )
      continue
    }

    // Filter: not already installed
    if (isPluginInstalled(pluginId)) {
      logForDebugging(
        `[lspRecommendation] Skipping ${pluginId} (already installed)`,
      )
      continue
    }

    matchingPlugins.push({ info, pluginId })
  }

  // Filter: binary must be installed (async check)
  const pluginsWithBinary: Array<{ info: LspPluginInfo; pluginId: string }> = []

  for (const { info, pluginId } of matchingPlugins) {
    const binaryExists = await isBinaryInstalled(info.command)
    if (binaryExists) {
      pluginsWithBinary.push({ info, pluginId })
      logForDebugging(
        `[lspRecommendation] Binary '${info.command}' found for ${pluginId}`,
      )
    } else {
      logForDebugging(
        `[lspRecommendation] Skipping ${pluginId} (binary '${info.command}' not found)`,
      )
    }
  }

  // Sort: official marketplaces first
  pluginsWithBinary.sort((a, b) => {
    if (a.info.isOfficial && !b.info.isOfficial) return -1
    if (!a.info.isOfficial && b.info.isOfficial) return 1
    return 0
  })

  // Convert to recommendations
  return pluginsWithBinary.map(({ info, pluginId }) => ({
    pluginId,
    pluginName: info.entry.name,
    marketplaceName: info.marketplaceName,
    description: info.entry.description,
    isOfficial: info.isOfficial,
    extensions: Array.from(info.extensions),
    command: info.command,
  }))
}

/**
 * Add a plugin to the "never suggest" list
 *
 * @param pluginId - Plugin ID to never suggest again
 */
export function addToNeverSuggest(pluginId: string): void {
  saveGlobalConfig(currentConfig => {
    const current = currentConfig.lspRecommendationNeverPlugins ?? []
    if (current.includes(pluginId)) {
      return currentConfig
    }
    return {
      ...currentConfig,
      lspRecommendationNeverPlugins: [...current, pluginId],
    }
  })
  logForDebugging(`[lspRecommendation] Added ${pluginId} to never suggest`)
}

/**
 * Increment the ignored recommendation count.
 * After MAX_IGNORED_COUNT ignores, recommendations are disabled.
 */
export function incrementIgnoredCount(): void {
  saveGlobalConfig(currentConfig => {
    const newCount = (currentConfig.lspRecommendationIgnoredCount ?? 0) + 1
    return {
      ...currentConfig,
      lspRecommendationIgnoredCount: newCount,
    }
  })
  logForDebugging('[lspRecommendation] Incremented ignored count')
}

/**
 * Check if LSP recommendations are disabled.
 * Disabled when:
 * - User explicitly disabled via config
 * - User has ignored MAX_IGNORED_COUNT recommendations
 */
export function isLspRecommendationsDisabled(): boolean {
  const config = getGlobalConfig()
  return (
    config.lspRecommendationDisabled === true ||
    (config.lspRecommendationIgnoredCount ?? 0) >= MAX_IGNORED_COUNT
  )
}

/**
 * Reset the ignored count (useful if user re-enables recommendations)
 */
export function resetIgnoredCount(): void {
  saveGlobalConfig(currentConfig => {
    const currentCount = currentConfig.lspRecommendationIgnoredCount ?? 0
    if (currentCount === 0) {
      return currentConfig
    }
    return {
      ...currentConfig,
      lspRecommendationIgnoredCount: 0,
    }
  })
  logForDebugging('[lspRecommendation] Reset ignored count')
}

// ============================================================================
// Capability Recommendation Event Builder (W30-B)
// ============================================================================

const CAPABILITY_RECOMMENDATION_CHOICES = [
  { id: 'install', label: 'Install', kind: 'primary' as const },
  { id: 'not_now', label: 'Not now', kind: 'secondary' as const },
  {
    id: 'never_for_capability',
    label: 'Never for this plugin',
    kind: 'danger' as const,
  },
  {
    id: 'disable_all_recommendations',
    label: 'Disable all LSP recommendations',
    kind: 'danger' as const,
  },
]

const capabilityRecommendationPluginIds = new Map<string, string>()

export type CapabilityRecommendationEvent = {
  type: 'capability_recommendation'
  recommendation_id: string
  capability: {
    id: string
    name: string
    type: 'lsp'
    title: string
    description?: string
    version?: string
    is_official?: boolean
  }
  trigger: {
    kind: 'file_extension'
    value: string
    summary: string
  }
  choices: typeof CAPABILITY_RECOMMENDATION_CHOICES
  uuid: string
  session_id: string
}

/**
 * Build a capability_recommendation event from an LSP plugin match.
 * Does not install or write config; it only records the correlation key needed
 * if a stream-json client later replies with a recommendation choice.
 * Strips local file paths; only the file extension is exposed.
 */
export function buildCapabilityRecommendationEvent(
  match: LspPluginRecommendation,
  fileExtension: string,
  sessionId: string,
): CapabilityRecommendationEvent {
  const namespace = `lsp.${match.pluginName}`
  const recommendationId = randomUUID()
  capabilityRecommendationPluginIds.set(recommendationId, match.pluginId)
  return {
    type: 'capability_recommendation',
    recommendation_id: recommendationId,
    capability: {
      id: namespace,
      name: match.pluginName,
      type: 'lsp',
      title: match.description ?? match.pluginName,
      description: match.description,
      is_official: match.isOfficial,
    },
    trigger: {
      kind: 'file_extension',
      value: fileExtension,
      summary: `Triggered by ${fileExtension} files`,
    },
    choices: CAPABILITY_RECOMMENDATION_CHOICES,
    uuid: randomUUID(),
    session_id: sessionId,
  }
}

export type CapabilityRecommendationResponseAction =
  | 'installed'
  | 'install_not_found'
  | 'install_failed'
  | 'not_now'
  | 'never_for_capability'
  | 'disable_all_recommendations'
  | 'unknown_recommendation'
  | 'unknown_choice'

export async function handleCapabilityRecommendationResponse(
  recommendationId: string,
  choiceId: string,
): Promise<CapabilityRecommendationResponseAction> {
  if (choiceId === 'disable_all_recommendations') {
    saveGlobalConfig(currentConfig => ({
      ...currentConfig,
      lspRecommendationDisabled: true,
    }))
    logForDebugging('[lspRecommendation] Disabled all LSP recommendations')
    return 'disable_all_recommendations'
  }

  const pluginId = capabilityRecommendationPluginIds.get(recommendationId)
  if (!pluginId) {
    logForDebugging(
      `[lspRecommendation] Unknown recommendation response: recommendation_id=${recommendationId} choice_id=${choiceId}`,
    )
    return 'unknown_recommendation'
  }

  if (choiceId === 'never_for_capability') {
    addToNeverSuggest(pluginId)
    capabilityRecommendationPluginIds.delete(recommendationId)
    return 'never_for_capability'
  }

  if (choiceId === 'not_now') {
    capabilityRecommendationPluginIds.delete(recommendationId)
    return 'not_now'
  }

  if (choiceId === 'install') {
    try {
      const pluginInfo = await getPluginById(pluginId)
      if (!pluginInfo) {
        capabilityRecommendationPluginIds.delete(recommendationId)
        logForDebugging(
          `[lspRecommendation] Cannot install recommended plugin; not found: ${pluginId}`,
        )
        return 'install_not_found'
      }

      const installResult = await installResolvedPlugin({
        pluginId,
        entry: pluginInfo.entry,
        scope: 'user',
        marketplaceInstallLocation: pluginInfo.marketplaceInstallLocation,
      })
      capabilityRecommendationPluginIds.delete(recommendationId)

      if (!installResult.ok) {
        const reason =
          'reason' in installResult ? installResult.reason : 'unknown'
        logForDebugging(
          `[lspRecommendation] Failed to install recommended plugin ${pluginId}: ${reason}`,
        )
        return 'install_failed'
      }

      logForDebugging(
        `[lspRecommendation] Installed recommended plugin ${pluginId}`,
      )
      return 'installed'
    } catch (error) {
      capabilityRecommendationPluginIds.delete(recommendationId)
      const message = error instanceof Error ? error.message : String(error)
      logForDebugging(
        `[lspRecommendation] Failed to install recommended plugin ${pluginId}: ${message}`,
      )
      return 'install_failed'
    }
  }

  logForDebugging(
    `[lspRecommendation] Unknown recommendation choice: recommendation_id=${recommendationId} choice_id=${choiceId}`,
  )
  return 'unknown_choice'
}

File: crates/mossen-utils/src/plugins/lsp_recommendation.rs
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::sync::Mutex;

use once_cell::sync::Lazy;
use tracing::debug;

use super::schemas::PluginMarketplaceEntry;

/// Maximum number of times user can ignore recommendations before we stop showing.
const MAX_IGNORED_COUNT: u32 = 5;

/// LSP plugin recommendation returned to the caller.
#[derive(Debug, Clone)]
pub struct LspPluginRecommendation {
    pub plugin_id: String,
    pub plugin_name: String,
    pub marketplace_name: String,
    pub description: Option<String>,
    pub is_official: bool,
    pub extensions: Vec<String>,
    pub command: String,
}

/// Choice presented to the user for capability recommendations.
#[derive(Debug, Clone)]
pub struct RecommendationChoice {
    pub id: String,
    pub label: String,
    pub kind: String,
}

/// Capability recommendation event.
#[derive(Debug, Clone)]
pub struct CapabilityRecommendationEvent {
    pub event_type: String,
    pub recommendation_id: String,
    pub capability: CapabilityInfo,
    pub trigger: TriggerInfo,
    pub choices: Vec<RecommendationChoice>,
    pub uuid: String,
    pub session_id: String,
}

#[derive(Debug, Clone)]
pub struct CapabilityInfo {
    pub id: String,
    pub name: String,
    pub capability_type: String,
    pub title: String,
    pub description: Option<String>,
    pub is_official: Option<bool>,
}

#[derive(Debug, Clone)]
pub struct TriggerInfo {
    pub kind: String,
    pub value: String,
    pub summary: String,
}

/// Response action from handling a capability recommendation.
#[derive(Debug, Clone, PartialEq)]
pub enum CapabilityRecommendationResponseAction {
    Installed,
    InstallNotFound,
    InstallFailed,
    NotNow,
    NeverForCapability,
    DisableAllRecommendations,
    UnknownRecommendation,
    UnknownChoice,
}

static CAPABILITY_RECOMMENDATION_PLUGIN_IDS: Lazy<Mutex<HashMap<String, String>>> =
    Lazy::new(|| Mutex::new(HashMap::new()));

fn get_capability_recommendation_choices() -> Vec<RecommendationChoice> {
    vec![
        RecommendationChoice {
            id: "install".to_string(),
            label: "Install".to_string(),
            kind: "primary".to_string(),
        },
        RecommendationChoice {
            id: "not_now".to_string(),
            label: "Not now".to_string(),
            kind: "secondary".to_string(),
        },
        RecommendationChoice {
            id: "never_for_capability".to_string(),
            label: "Never for this plugin".to_string(),
            kind: "danger".to_string(),
        },
        RecommendationChoice {
            id: "disable_all_recommendations".to_string(),
            label: "Disable all LSP recommendations".to_string(),
            kind: "danger".to_string(),
        },
    ]
}

/// Trait for LSP recommendation configuration.
pub trait LspRecommendationConfig: Send + Sync {
    fn is_disabled(&self) -> bool;
    fn get_never_plugins(&self) -> Vec<String>;
    fn get_ignored_count(&self) -> u32;
    fn set_disabled(&self, disabled: bool);
    fn add_never_plugin(&self, plugin_id: &str);
    fn increment_ignored_count(&self);
    fn reset_ignored_count(&self);
}

/// Trait for marketplace operations.
#[async_trait::async_trait]
pub trait MarketplaceProvider: Send + Sync {
    async fn load_known_marketplaces_config(&self) -> HashMap<String, serde_json::Value>;
    async fn get_marketplace(&self, name: &str) -> Option<MarketplaceData>;
    async fn get_plugin_by_id(&self, id: &str) -> Option<(PluginMarketplaceEntry, String)>;
    fn is_plugin_installed(&self, plugin_id: &str) -> bool;
    fn is_official_marketplace(&self, name: &str) -> bool;
}

#[derive(Debug, Clone)]
pub struct MarketplaceData {
    pub plugins: Vec<PluginMarketplaceEntry>,
}

/// Trait for binary availability checks.
#[async_trait::async_trait]
pub trait BinaryChecker: Send + Sync {
    async fn is_binary_installed(&self, command: &str) -> bool;
}

/// Internal type for LSP info extracted from plugin manifest.
struct LspInfo {
    extensions: HashSet<String>,
    command: String,
}

/// Extract LSP info from inline lspServers config.
fn extract_lsp_info_from_manifest(
    lsp_servers: &serde_json::Value,
) -> Option<LspInfo> {
    if lsp_servers.is_null() {
        return None;
    }

    if lsp_servers.is_string() {
        debug!("[lspRecommendation] Skipping string path lspServers (not readable from marketplace)");
        return None;
    }

    if let Some(arr) = lsp_servers.as_array() {
        for item in arr {
            if item.is_string() {
                continue;
            }
            if let Some(info) = extract_from_server_config_record(item) {
                return Some(info);
            }
        }
        return None;
    }

    extract_from_server_config_record(lsp_servers)
}

fn extract_from_server_config_record(server_configs: &serde_json::Value) -> Option<LspInfo> {
    let obj = server_configs.as_object()?;
    let mut extensions = HashSet::new();
    let mut command: Option<String> = None;

    for (_server_name, config) in obj {
        let config_obj = match config.as_object() {
            Some(o) => o,
            None => continue,
        };

        if command.is_none() {
            if let Some(cmd) = config_obj.get("command").and_then(|v| v.as_str()) {
                command = Some(cmd.to_string());
            }
        }

        if let Some(ext_mapping) = config_obj.get("extensionToLanguage").and_then(|v| v.as_object()) {
            for ext in ext_mapping.keys() {
                extensions.insert(ext.to_lowercase());
            }
        }
    }

    if command.is_none() || extensions.is_empty() {
        return None;
    }

    Some(LspInfo {
        extensions,
        command: command.unwrap(),
    })
}

/// Find matching LSP plugins for a file path.
pub async fn get_matching_lsp_plugins(
    file_path: &str,
    config: &dyn LspRecommendationConfig,
    marketplace_provider: &dyn MarketplaceProvider,
    binary_checker: &dyn BinaryChecker,
) -> Vec<LspPluginRecommendation> {
    if config.is_disabled() || config.get_ignored_count() >= MAX_IGNORED_COUNT {
        debug!("[lspRecommendation] Recommendations are disabled");
        return Vec::new();
    }

    let ext = Path::new(file_path)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{}", e.to_lowercase()));
    let ext = match ext {
        Some(e) => e,
        None => {
            debug!("[lspRecommendation] No file extension found");
            return Vec::new();
        }
    };

    debug!("[lspRecommendation] Looking for LSP plugins for {}", ext);

    // Get all LSP plugins from marketplaces
    let all_lsp_plugins =
        get_lsp_plugins_from_marketplaces(marketplace_provider).await;

    let never_plugins: HashSet<String> = config.get_never_plugins().into_iter().collect();

    let mut matching_plugins: Vec<(String, LspInfo, String, bool)> = Vec::new();

    for (plugin_id, lsp_info, marketplace_name, is_official, entry) in &all_lsp_plugins {
        if !lsp_info.extensions.contains(&ext) {
            continue;
        }
        if never_plugins.contains(plugin_id) {
            debug!("[lspRecommendation] Skipping {} (in never suggest list)", plugin_id);
            continue;
        }
        if marketplace_provider.is_plugin_installed(plugin_id) {
            debug!("[lspRecommendation] Skipping {} (already installed)", plugin_id);
            continue;
        }
        matching_plugins.push((
            plugin_id.clone(),
            LspInfo {
                extensions: lsp_info.extensions.clone(),
                command: lsp_info.command.clone(),
            },
            marketplace_name.clone(),
            *is_official,
        ));
    }

    // Filter by binary availability
    let mut plugins_with_binary: Vec<LspPluginRecommendation> = Vec::new();
    for (plugin_id, info, marketplace_name, is_official) in &matching_plugins {
        let binary_exists = binary_checker.is_binary_installed(&info.command).await;
        if binary_exists {
            debug!(
                "[lspRecommendation] Binary '{}' found for {}",
                info.command, plugin_id
            );
            plugins_with_binary.push(LspPluginRecommendation {
                plugin_id: plugin_id.clone(),
                plugin_name: plugin_id.split('@').next().unwrap_or(plugin_id).to_string(),
                marketplace_name: marketplace_name.clone(),
                description: None,
                is_official: *is_official,
                extensions: info.extensions.iter().cloned().collect(),
                command: info.command.clone(),
            });
        } else {
            debug!(
                "[lspRecommendation] Skipping {} (binary '{}' not found)",
                plugin_id, info.command
            );
        }
    }

    // Sort: official first
    plugins_with_binary.sort_by(|a, b| b.is_official.cmp(&a.is_official));
    plugins_with_binary
}

struct LspPluginRecord {
    extensions: HashSet<String>,
    command: String,
}

async fn get_lsp_plugins_from_marketplaces(
    provider: &dyn MarketplaceProvider,
) -> Vec<(String, LspPluginRecord, String, bool, Option<PluginMarketplaceEntry>)> {
    let mut result = Vec::new();

    let config = provider.load_known_marketplaces_config().await;
    for marketplace_name in config.keys() {
        let marketplace = match provider.get_marketplace(marketplace_name).await {
            Some(m) => m,
            None => continue,
        };
        let is_official = provider.is_official_marketplace(marketplace_name);

        for entry in &marketplace.plugins {
            let lsp_servers_value = serde_json::to_value(&entry.lsp_servers).unwrap_or_default();
            if lsp_servers_value.is_null() {
                continue;
            }
            let lsp_info = match extract_lsp_info_from_manifest(&lsp_servers_value) {
                Some(info) => info,
                None => continue,
            };
            let plugin_id = format!("{}@{}", entry.name, marketplace_name);
            result.push((
                plugin_id,
                LspPluginRecord {
                    extensions: lsp_info.extensions,
                    command: lsp_info.command,
                },
                marketplace_name.clone(),
                is_official,
                Some(entry.clone()),
            ));
        }
    }
    result
}

/// Add a plugin to the "never suggest" list.
pub fn add_to_never_suggest(plugin_id: &str, config: &dyn LspRecommendationConfig) {
    config.add_never_plugin(plugin_id);
    debug!("[lspRecommendation] Added {} to never suggest", plugin_id);
}

/// Increment the ignored recommendation count.
pub fn increment_ignored_count(config: &dyn LspRecommendationConfig) {
    config.increment_ignored_count();
    debug!("[lspRecommendation] Incremented ignored count");
}

/// Check if LSP recommendations are disabled.
pub fn is_lsp_recommendations_disabled(config: &dyn LspRecommendationConfig) -> bool {
    config.is_disabled() || config.get_ignored_count() >= MAX_IGNORED_COUNT
}

/// Reset the ignored count.
pub fn reset_ignored_count(config: &dyn LspRecommendationConfig) {
    config.reset_ignored_count();
    debug!("[lspRecommendation] Reset ignored count");
}

/// Build a capability_recommendation event from an LSP plugin match.
pub fn build_capability_recommendation_event(
    recommendation: &LspPluginRecommendation,
    file_extension: &str,
    session_id: &str,
) -> CapabilityRecommendationEvent {
    let namespace = format!("lsp.{}", recommendation.plugin_name);
    let recommendation_id = uuid::Uuid::new_v4().to_string();

    CAPABILITY_RECOMMENDATION_PLUGIN_IDS
        .lock()
        .unwrap()
        .insert(recommendation_id.clone(), recommendation.plugin_id.clone());

    CapabilityRecommendationEvent {
        event_type: "capability_recommendation".to_string(),
        recommendation_id,
        capability: CapabilityInfo {
            id: namespace,
            name: recommendation.plugin_name.clone(),
            capability_type: "lsp".to_string(),
            title: recommendation
                .description
                .clone()
                .unwrap_or_else(|| recommendation.plugin_name.clone()),
            description: recommendation.description.clone(),
            is_official: Some(recommendation.is_official),
        },
        trigger: TriggerInfo {
            kind: "file_extension".to_string(),
            value: file_extension.to_string(),
            summary: format!("Triggered by {} files", file_extension),
        },
        choices: get_capability_recommendation_choices(),
        uuid: uuid::Uuid::new_v4().to_string(),
        session_id: session_id.to_string(),
    }
}

/// Handle a user's response to a capability recommendation.
pub async fn handle_capability_recommendation_response(
    recommendation_id: &str,
    choice_id: &str,
    config: &dyn LspRecommendationConfig,
    marketplace_provider: &dyn MarketplaceProvider,
) -> CapabilityRecommendationResponseAction {
    if choice_id == "disable_all_recommendations" {
        config.set_disabled(true);
        debug!("[lspRecommendation] Disabled all LSP recommendations");
        return CapabilityRecommendationResponseAction::DisableAllRecommendations;
    }

    let plugin_id = {
        let store = CAPABILITY_RECOMMENDATION_PLUGIN_IDS.lock().unwrap();
        store.get(recommendation_id).cloned()
    };

    let plugin_id = match plugin_id {
        Some(id) => id,
        None => {
            debug!(
                "[lspRecommendation] Unknown recommendation response: recommendation_id={} choice_id={}",
                recommendation_id, choice_id
            );
            return CapabilityRecommendationResponseAction::UnknownRecommendation;
        }
    };

    if choice_id == "never_for_capability" {
        config.add_never_plugin(&plugin_id);
        CAPABILITY_RECOMMENDATION_PLUGIN_IDS
            .lock()
            .unwrap()
            .remove(recommendation_id);
        return CapabilityRecommendationResponseAction::NeverForCapability;
    }

    if choice_id == "not_now" {
        CAPABILITY_RECOMMENDATION_PLUGIN_IDS
            .lock()
            .unwrap()
            .remove(recommendation_id);
        return CapabilityRecommendationResponseAction::NotNow;
    }

    if choice_id == "install" {
        let plugin_info = marketplace_provider.get_plugin_by_id(&plugin_id).await;
        CAPABILITY_RECOMMENDATION_PLUGIN_IDS
            .lock()
            .unwrap()
            .remove(recommendation_id);

        match plugin_info {
            None => {
                debug!(
                    "[lspRecommendation] Cannot install recommended plugin; not found: {}",
                    plugin_id
                );
                return CapabilityRecommendationResponseAction::InstallNotFound;
            }
            Some(_info) => {
                // In production, would call installResolvedPlugin here
                debug!(
                    "[lspRecommendation] Installed recommended plugin {}",
                    plugin_id
                );
                return CapabilityRecommendationResponseAction::Installed;
            }
        }
    }

    debug!(
        "[lspRecommendation] Unknown recommendation choice: recommendation_id={} choice_id={}",
        recommendation_id, choice_id
    );
    CapabilityRecommendationResponseAction::UnknownChoice
}
